### ISSUE TITLE
CancelledError and misbehaving clients handling

### ISSUE THREAD

**Issue by washeck** [aio-libs/aiosmtpd#110]
I tried to write a RCPT handler that accepts connection from various embedded devices and forwards the email address to an external system. However, as the devices are not proper SMTP clients and their quality varies I need the solution to be pretty robust. While developing it I discovered a problem with the way aoismtpd deals with clients who send data and ignore the response. To reproduce:

sleep.py (simple handler class simulating my handler):

```
import asyncio
from aiosmtpd.handlers import Debugging

class RcptWebhook(Debugging):
    async def handle_RCPT(self, server, session, envelope, address, rcpt_options):
        await asyncio.sleep(10)

        envelope.rcpt_tos.append(address)
        return '250 OK'
```

Run aiosmtpd

    python3 -m aiosmtpd -n -c sleep.RcptWebhook

smtp.txt (sample SMTP session)

```
HELO there
MAIL FROM: somebody@example.com
RCPT TO: anybody@example.com
DATA
test
.
```
(note the is no QUIT in the data)

Send it witch netcat

    nc localhost 8025 < smtp.txt

It sends the content of the file and EOF which causes unhandled CanceledError in aiosmtpd while it is waiting for my sleep handler to finish:
```

INFO:mail.log:Server listening on localhost:8025
INFO:mail.log:Starting asyncio loop
INFO:mail.log:Peer: ('127.0.0.1', 39306)
INFO:mail.log:('127.0.0.1', 39306) handling connection
INFO:mail.log:('127.0.0.1', 39306) Data: b'HELO there'
INFO:mail.log:('127.0.0.1', 39306) Data: b'MAIL FROM: somebody@example.com'
INFO:mail.log:('127.0.0.1', 39306) sender: somebody@example.com
INFO:mail.log:('127.0.0.1', 39306) Data: b'RCPT TO: anybody@example.com'
INFO:mail.log:('127.0.0.1', 39306) EOF received
ERROR:mail.log:SMTP session exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/aiosmtpd/smtp.py", line 276, in _handle_client
    yield from method(arg)
  File "/usr/local/lib/python3.5/dist-packages/aiosmtpd/smtp.py", line 582, in smtp_RCPT
    'RCPT', address, rcpt_options)
  File "/usr/local/lib/python3.5/dist-packages/aiosmtpd/smtp.py", line 116, in _call_handler_hook
    status = yield from hook(self, self.session, self.envelope, *args)
  File "/home/vaclav/src/smtp-events-receiver/sleep.py", line 6, in handle_RCPT
    await asyncio.sleep(10)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 516, in sleep
    return (yield from future)
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 266, in result
    raise CancelledError
concurrent.futures._base.CancelledError
INFO:mail.log:('127.0.0.1', 39306) Data: b'DATA'
INFO:mail.log:('127.0.0.1', 39306) Data: b'test'
INFO:mail.log:('127.0.0.1', 39306) Data: b'.'
INFO:mail.log:('127.0.0.1', 39306) Data: b''
INFO:mail.log:('127.0.0.1', 39306) Data: b''
INFO:mail.log:('127.0.0.1', 39306) Data: b''
INFO:mail.log:('127.0.0.1', 39306) Data: b''
...
```
aiosmtpd is now 100% busy on CPU and keeps printing "500 Error: bad syntax" to the netcat. It ignores KeyboardInterrupt and to quit it I need to terminate netcat with ^C which stops aiosmtp with "Connection lost".

I think aiosmtpd should handle this more gracefully as you don't want your server to get stuck on CPU due to misbehaving client. I see in RFC 5321 that the server must not close the connection until QUIT from the client so I'm not sure what is the proper way of handling EOF though.

**Comment by Mortal** [comment]
My patch in #59 should hopefully fix this.

Unfortunately I don't think I will have the time to follow up on that PR until after June 20th.

**Comment by warsaw** [comment]
@Mortal Do you think you'd be able to refactor out from #59 just the fix and a test for the problem reported here?  No worries if it's after the 20th.

**Comment by washeck** [comment]
@Mortal I tried #59 and confirm it fixes my problem.

**Comment by Mortal** [comment]
@warsaw Sure, I'll rebase and split that PR when I have the time. It's good to know that it fixes @washeck's problem.

**Comment by Mortal** [comment]
I will look at this after the 23rd, since my deadline has been extended by
three days.

**Comment by Mortal** [comment]
I've opened #115. @washeck will you try out that PR and let me know if it fixes your problem? I ended up finding a much simpler fix than what's in #59.

**Comment by washeck** [comment]
@Mortal Yes, your PR fixes my problem. Thanks for such a quick fix.
